app/services/weather.py

```diff
@@ -0,0 +1,329 @@
+import asyncio
+import json
+import logging
+import time
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import Any, Dict, List, Optional, Tuple
+
+import aiohttp
+
+from app.config.loader import ConfigLoader
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class NormalizedForecast:
+    """Lightweight, normalized daily forecast representation in Celsius.
+
+    This data model keeps only the fields that are useful for itinerary planning
+    and for concise UI display. Temperatures are expected in Celsius.
+    """
+
+    date: str
+    label: str  # e.g., "sunny", "rainy", "cloudy", "snowy", "cold", "hot"
+    temp_min_c: float
+    temp_max_c: float
+    precipitation_mm: float
+    wind_mps: float
+    description: str
+
+
+class _TTLCache:
+    """Simple in-memory TTL cache for storing forecasts per (city, start, end).
+
+    Note:
+        - This is process-local and ephemeral. It's sufficient for a single-process
+          Streamlit/CLI run. For multi-process deployments, a shared cache should be used.
+    """
+
+    def __init__(self, ttl_seconds: int) -> None:
+        self._ttl = max(0, int(ttl_seconds))
+        self._store: Dict[str, Tuple[float, Any]] = {}
+
+    def _key(self, city: str, start: str, end: str) -> str:
+        return json.dumps([city.strip().lower(), start, end])
+
+    def get(self, city: str, start: str, end: str) -> Optional[Any]:
+        if self._ttl == 0:
+            return None
+        key = self._key(city, start, end)
+        entry = self._store.get(key)
+        if not entry:
+            return None
+        ts, value = entry
+        if time.time() - ts > self._ttl:
+            self._store.pop(key, None)
+            return None
+        return value
+
+    def set(self, city: str, start: str, end: str, value: Any) -> None:
+        if self._ttl == 0:
+            return
+        key = self._key(city, start, end)
+        self._store[key] = (time.time(), value)
+
+
+# Process-wide cache instance (initialized on first WeatherService creation)
+_GLOBAL_CACHE: Optional[_TTLCache] = None
+
+
+def _coerce_date(d: Any) -> datetime:
+    if isinstance(d, datetime):
+        return d
+    if isinstance(d, (int, float)):
+        return datetime.fromtimestamp(d)
+    if isinstance(d, str):
+        # Accept common formats; fallback to ISO
+        for fmt in ("%Y-%m-%d", "%Y/%m/%d", "%d-%m-%Y", "%d/%m/%Y"):
+            try:
+                return datetime.strptime(d, fmt)
+            except ValueError:
+                pass
+        return datetime.fromisoformat(d)
```

Comment:
The fallback to datetime.fromisoformat may throw ValueError for non-ISO formats. Add a try/except block for safer handling.
```
try:
    return datetime.fromisoformat(d)
except Exception:
    raise ValueError(f"Unsupported date value: {d}")
```